readme.md

```diff
@@ -0,0 +1,183 @@
+
+### Go sockets
+
+When we invoke the `net.Listen()` function in Go, this function use the `ListenConfig` struct to create the Listener.
+
+```go
+func Listen(network, address string) (Listener, error) {
+	var lc ListenConfig
+	return lc.Listen(context.Background(), network, address)
+}
+```
+
+If we inspect `ListenConfig,` we can find the method `Control,` and the documentation says: <mark>If Control is not nil, it is called after creating the network connection but before binding it to the operating system.</mark>
+
+```go
+// ListenConfig contains options for listening to an address.
+type ListenConfig struct {
+	// If Control is not nil, it is called after creating the network
+	// connection but before binding it to the operating system.
+	// ...
+	Control func(network, address string, c syscall.RawConn) error
+	...
+}
+```
+
+The function `Control` receives the `syscall.RawConn` which is a raw network connection that has a method also called control (`Control(f func(fd uintptr))`) where it will invoke the function `f` on the underlying connection's file descriptor.
+
+Having the file descriptor we can use the `golang.org/x/sys/unix` package to set the socket options.
+
+```go
+// fd -> the underlying connection's file descriptor.
+// unix.SOL_SOCKET  -> to set options at the socket level, we have to specify the level argument as SOL_SOCKET.
+unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+```
+
+So, we could create a instance of `ListenConfig` with a control function to set the `SO_REUSEPORT` socket options to our sockets.
+
+```go
+var lc = net.ListenConfig{
+	Control: func(network, address string, c syscall.RawConn) error {
+		var opErr error
+		if err := c.Control(func(fd uintptr) {
+			opErr = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+		}); err != nil {
+			return err
+		}
+		return opErr
+	},
+}
+```
+
+### Security
+
+One question we might have at this point is, what about security? I mean, if we can open a socket with the same IP: Port of a specific app, for example, Nginx, we could hijack part of the requests that the kernel will send to us through the socket. Right?v
+
+Well, to prevent this "port hijacking," Linux has special limitations or mechanisms such as:
+
+* Both sockets must have been created with the SO_REUSEPORT socket option. If there is a socket running without SO_REUSEPORT and we try to create another socket even with the SO_REUSEPORT socket option, it will fail with the error `already in use`. 
+* All sockets that want to listen to the same IP and port combination must have the same effective userID. For example, if you want to hijack the Nginx port and it is running under the ownership of the user Pepito, a new process can listen to the same port only if it is also owned by the user Pepito. So one user cannot "steal" ports of other users.
```

Comment:
Esto es un poco más técnico, pero Docker containers no tienen usuarios (es decir, tienen user IDs pero no usuarios como tal). Es posible mapear un usuario interno a uno externo si conoces el `id -u` and `id -g`, pero dentro del container, el usuario será `nobody` a menos que sea agregado a `/etc/passwd`. 

Entonces la pregunta es, sabiendo eso, es el ownership de `SO_REUSEPORT` via "nombre de usuario" o UID:GID combo? Yo creo que es el último, pero quizás esté equivocado.